{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Linear Classifier for Clear-Cell Renal Carcinoma Grade\n",
    "\n",
    "### Background:\n",
    "Clear-cell renal cell carcinoma (ccRCC) is the most common subtype of renal cancers, and accounts for nearly 50,000 new cases annually. The Cancer Genome Atlas (TCGA) project profiled over 400 sporadic ccRCC tumors, and published dozens of clinical and molecular features describing these tumors. My lab has studied 40 tumors from heritable cases of ccRCC. Individuals with heritable ccRCC have a germline loss-of-function mutation in the gene VHL, and make up nearly 2% of all ccRCC cases. We have observed differences in the number of somatic genetic driver events within high and low Fuhrman grade tumors. The Fuhrman grading system is assigned by pathologist based on the shape and size of the nucleus of tumor cells, and classifies tumors into one of four possible grades. High Fuhrman grade ccRCC tumors have an increased ability for metastasis and lower survival rate. There is very limited information regarding the clinical and molecular features of high and low Fuhrman grade ccRCCs. The goal of my project will be to build a classifier for Fuhrman grade in ccRCCs.\n",
    "\n",
    "### Summary of Features \n",
    "| **ID** | **Type** | **Description** |\n",
    "|:-------------:|:-------------:|:-----:|\n",
    "| Grade | Binary | Class variable. 0 = low Fuhrman grade, 1 = high Fuhrman grade |\n",
    "| Gender | Binary | Male = 0, Female = 1 |\n",
    "| Age | Discrete | Age at diagnosis |\n",
    "| MaxDim | Continuous | Maximum tumor dimension, surrogate for tumor size |\n",
    "| Mets | Binary | Presence of metastasis = 1, absence of metastasis = 0 |\n",
    "| VHL | Binary | Presence of somatic mutation in VHL =1, absence = 0 |\n",
    "| Stage | Discrete | Tumor stages. 0 = stage 1, 1 = stage 2, 2 = stage 3, 3 = stage 4. |\n",
    "| *TotalMut* | Discrete | Total number of somatic exonic mutations |\n",
    "| *TotalNonSyn* | Discrete | Total number of somatic exonic nonsynonymous mutations |\n",
    "| *TotalSNVDriver* | Discrete | Total number of somatic exonic nonsynonymous mutations in known cancer driver genes |\n",
    "| *ClonSNVDriver* | Discrete | Total number of somatic exonic nonsynonymous clonal driver mutations in know cancer driver genes |\n",
    "| *SubclonSNVDriver* | Discrete | Total number of somatic exonic nonsynonymous subclonal driver mutations in know cancer driver genes |\n",
    "| *DriverSNVRatio* | Continuous | Number of driver mutations / total number of exonic nonsynonymous mutations |\n",
    "| *ClonSNVRatio* | Continuous | Number of clonal driver mutations / total number of driver mutations |\n",
    "\n",
    "##### Low Grade = 196 cases\n",
    "##### High Grade = 249 cases\n",
    "\n",
    "### Methods:\n",
    "Italics denotes features I computed for this project, and all other features are publicly available from TCGA. I will apply linear discriminant analysis on my feature set, and determine which (if any) features are linearly separable for each model. I will implement both learning algorithms from scratch. I will use leave-one-out cross-validation to determine how well each model generalizes. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions for calculating italicized features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Tue Oct 20 08:13:58 2015\n",
    "\n",
    "@author: mitcheas@ohsum01.ohsu.edu\n",
    "\n",
    "Calculates the cancer cell fraction for variants in a given VCF\n",
    "\n",
    "Usage: CCF.py -vcf <input.vcf> -seg <input.seg> -pur <0.75> -o\n",
    "Required inputs: \n",
    "1. vcf - VCF file, tumor read counts must be listed in the second to last column\n",
    "2. seg  -  SEG file containing copy number ratio in log base 2 \n",
    "3. pur - a value between 0 and 1 representing tumor purity\n",
    "4. o - path to your output file\n",
    "\n",
    "Output file will be written to the filename provided at input\n",
    "The output file is a tab delimited file with the following headers:\n",
    "\n",
    "CHROM\tPOS\tREAD_DEPTH\tCCF\tERROR\tCLONAL\n",
    "\n",
    "CHROM - the chromosome (ie 1, 2, ..., X)\n",
    "POS - nucleotide position\n",
    "READ_DEPTH - number of reads mapped to nucleotide position\n",
    "CCF - cancer cell fraction (usually a value between 0 and 1, values above 1 are likely and\n",
    "\t\trepresent errors in purity, allele frequency, and integer copy number estimation)\n",
    "ERROR - +/- error for the 95% confidence interval\n",
    "CLONAL - binary T/F where T denotes clonal SNV and F denotes a subclonal SNV\n",
    "\"\"\"\n",
    "\n",
    "import math\n",
    "import numpy as np\n",
    "import sys\n",
    "import string, re\n",
    "from scipy import stats\n",
    "import argparse\n",
    "\n",
    "np.seterr(divide='ignore', invalid='ignore')\n",
    "\n",
    "nonsyn_mut = ['frameshift_variant', 'missense_variant', 'splice_acceptor_variant',\n",
    "\t\t\t\t'splice_region_variant', 'stop_gained', 'inframe_deletion']\n",
    "\n",
    "# FUNCTION: READ_FILE(FILENAME)\n",
    "# Opens and Reads all lines of an input file \n",
    "def read_file(Filename):\n",
    "\tFilename = open(Filename, 'U')\n",
    "\talllines = Filename.readlines()\n",
    "\toutlist = []\n",
    "\tfor line in alllines:\n",
    "\t\tline = line.rstrip()\n",
    "\t\tline = string.split(line, '\\t')\n",
    "\t\toutlist.append(line)\n",
    "\treturn outlist\n",
    "    \n",
    "# FUNCTION: READ_FILE_hd(FILENAME)\n",
    "# Opens and Reads all lines of an input file with a header\n",
    "def read_file_hd(Filename):\n",
    "\tFilename = open(Filename, 'U')\n",
    "\talllines = Filename.readlines()\n",
    "\talllines = alllines[1:len(alllines)]\n",
    "\toutlist = []\n",
    "\tfor line in alllines:\n",
    "\t\tline = line.rstrip()\n",
    "\t\tline = string.split(line, '\\t')\n",
    "\t\toutlist.append(line)\n",
    "\treturn outlist\n",
    "\n",
    "# FUNCTION: READ_VCF(FILENAME)\n",
    "# Reads all lines of an input VCF file\n",
    "# Removes header lines of VCF\n",
    "def read_vcf(Filename):\n",
    "\tFilename = open(Filename, 'U')\n",
    "\talllines = Filename.readlines()\n",
    "\toutlist = []\n",
    "\tfor line in alllines:\n",
    "\t\tif re.match('##SAMPLE=<ID=TUMOR', line):\n",
    "\t\t\ttumor = string.split(line, ',')[1]\n",
    "\t\t\ttumor = string.split(tumor, '=')[1]\n",
    "\t\t\ttumor = string.join((string.split(tumor, '-')[0:3]), '-')\n",
    "\t\telif not re.match('#', line):\n",
    "\t\t\tline = line.rstrip()\n",
    "\t\t\tline = string.split(line, '\\t')\n",
    "\t\t\toutlist.append(line)\n",
    "\treturn tumor, outlist\n",
    "\n",
    "# FUNCTION: CCF(VAF, Purity, Ploidy, CN)\n",
    "# Calculate the cancer cell fraction\n",
    "def CCF(mreads, wtreads, pur, cnT, cnN):\n",
    "\ttreads = float(mreads) + float(wtreads)\n",
    "\tvaf = float(mreads)/float(treads)\n",
    "\tcnT = float(cnT)  \n",
    "\tn_obs = float(vaf/pur) * ((pur * cnT) + (cnN * (1-pur)))\n",
    "\tn_exp = Nexp(int(cnT))\n",
    "\tx = n_obs/n_exp\n",
    "\treturn x\n",
    "\n",
    "# FUNCTION: Nexp(CN):\n",
    "# CN = integer copy number returned from COPY_NUM function\n",
    "# Calculate and return the expected mutation copy number, n\n",
    "# Expected mutation copy number is estimated by determining the maximum likelihood\n",
    "# of the copy number state given the observed copy number and the assumption that\n",
    "# a somatic mutation is most likely to be heterozygous.\n",
    "#\n",
    "# To calculate Nexp without the assumption that a somatic mutation is heterozygous (ie in \n",
    "# the case of mutations residing in oncogenes), you can edit this function to also require\n",
    "# the VAF (variant allele frequency) of a mutation.\n",
    "def Nexp(cn):\n",
    "\tsuccess = np.zeros(cn)\n",
    "\ttotal = np.zeros(cn)\n",
    "\tcn_range = np.arange(1, cn+1)\n",
    "\tfor i in range(1,cn+1):\n",
    "\t\tx = float(math.factorial(cn))\n",
    "\t\tx = x/(float(math.factorial(i))*float(math.factorial(cn-i)))\n",
    "\t\ttotal[i-1] = x\n",
    "\tsuccess = total * cn_range\n",
    "\tn = sum(success)/ sum(total)\n",
    "\treturn n\n",
    "\n",
    "#FUNCTION: bootstrap_CCF(mreads, wtreads)\n",
    "# Bootstrap resampling of the mutant (mreads) and wild-type (wtreads) reads to create a \n",
    "# distribution of possible mutant/wild-type read populations. Determines CCF for every\n",
    "# bootstrap sample. Calculates and returns the error (95% confidence interval of the ccf)\n",
    "def Bootstrap_CCF(mreads, wtreads, pur, cnT, cnN):\n",
    "\tI = 10000\n",
    "\tccf = []\n",
    "\tfor x in range(I):\n",
    "\t\tm, wt = Resample(mreads, wtreads)\n",
    "\t\tc = CCF(m, wt, pur, cnT, cnN)\n",
    "\t\tccf.append(c)\n",
    "\tci = ConfInv(ccf)\n",
    "\treturn ci\n",
    "\t\t\n",
    "# Calculate and return the error based on a 95% confidence interval\n",
    "def ConfInv(ccf):\n",
    "\tn, min_max, mean, var, skew, kurt = stats.describe(ccf)\n",
    "\ts = math.sqrt(var)\n",
    "\tc = float(s) * float(1.96)\n",
    "\treturn c\n",
    "\n",
    "# Return a count for mutant and wild-type reads, from a random resample. \n",
    "def Resample(mreads, wtreads):\n",
    "\treads = np.concatenate((np.ones(mreads), np.zeros(wtreads)))\n",
    "\tx = np.random.choice(reads, (mreads+wtreads), replace=\"TRUE\")\n",
    "\tm = sum(x)\n",
    "\twt = len(x) - m\n",
    "\treturn m, wt\n",
    "    \n",
    "# FUNCTION:  COPYNUM(seg_dict, chrm, pos, cnN)\n",
    "# Calculate and return the integer copy number from a SEG file for a nucleotide position \n",
    "# on a given chromosome, where the average tumor copy number ratio is provided in log base \n",
    "# 2 format. Seg_dict is a dictionary containing all log base 2 copy number ratios read from\n",
    "# the seg file.\n",
    "def Copy_Num(seg_dict, chrm, pos, cnN):\n",
    "    d = []\n",
    "    if chrm not in seg_dict.keys():\n",
    "\t\tc = 2.0\n",
    "\t\treturn c\n",
    "    elif pos < sorted(seg_dict[chrm].keys())[0]:\n",
    "\t\tc = 2.0\n",
    "\t\treturn c\n",
    "    else:\n",
    "        for start in sorted(seg_dict[chrm].keys()):\n",
    "            if pos >= start:\n",
    "            \td.append(start)\n",
    "        for start in d:\n",
    "\t\t\tend = seg_dict[chrm][start][0]\n",
    "\t\t\tif pos <= (end+1):\n",
    "\t\t\t\tcn = seg_dict[chrm][start][1]\n",
    "\t\t\t\tc1 = pow(2, cn)\n",
    "\t\t\t\tc = c1 * cnN\n",
    "\t\t\t\tc = round(c)\n",
    "\t\t\t\treturn c\n",
    "\t\t\telse:\n",
    "\t\t\t\tc = 2.0\n",
    "    return c\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Main code for calculating features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "\t\n",
    "\tFEATURES = {}\n",
    "\tTMP = read_file_hd('KIRC_features.txt')\n",
    "\tfor line in TMP:\n",
    "\t\tt_id = line[0]\n",
    "\t\tgrade = line[1]\n",
    "\t\tgender = line[2]\n",
    "\t\tage = line[3]\n",
    "\t\tmaxd = line[4]\n",
    "\t\tmets = =line[5]\n",
    "\t\tstage = line[6]\n",
    "\t\tif t_id not in FEATURES.keys():\n",
    "\t\t\tFEATURES[t_id] = []\n",
    "\t\tFEATURES[t_id] = [grade, gender, age, mxd, mets, stage]\n",
    "\t\t\n",
    "\tBIC_SEG = {} ## dictionary for copy number\n",
    "\tTMP = read_file_hd('KIRC_edited.seg')\n",
    "\tfor line in TMP:\n",
    "\t\tt_id = string.join((string.split(line[0], '-')[0:3]), '-')\n",
    "\t\tchr = line[1]\n",
    "\t\tst = int(line[2])\n",
    "\t\tend = int(line[3])\n",
    "\t\tseg = float(line[5])\n",
    "\t\tif t_id not in BIC_SEG.keys():\n",
    "\t\t\tBIC_SEG[t_id] = {}\n",
    "\t\tif chr not in BIC_SEG[t_id].keys():\n",
    "\t\t\tBIC_SEG[t_id][chr] = {}\n",
    "\t\tif st not in BIC_SEG[t_id][chr].keys():\n",
    "\t\t\tBIC_SEG[t_id][chr][st] = []\n",
    "\t\tBIC_SEG[t_id][chr][st] = [end, seg]\n",
    "\t\n",
    "\tDRIVERS = read_file('drivers.txt')\n",
    "\t\n",
    "\t## Loop through all VCF files\t\n",
    "\tinfiles = read_file('vcffiles.txt')\n",
    "\t\n",
    "\tfor fi in infiles:\n",
    "\t\t# Open and read input files\n",
    "\t\ttum_id, vcf = read_vcf(fi)\t\t\n",
    "\t\t\n",
    "\t\t# Check that tumor ID is in features list and BIC_SEG\n",
    "\t\tif tum_id in FEATURES.keys() and tum_id in BIC_SEG.keys():\n",
    "\t\t\tgender = FEATURES[tum_id][1]\n",
    "\t\t\t# Write header to output file\n",
    "\t\t\t# outfile.write('CHROM\\tPOS\\tREAD_DEPTH\\tCCF\\tERROR\\tCLONAL\\n')\n",
    "\n",
    "\t\t\ttmp = [0, 0, 0, 0, 0, 0, 0]\n",
    "\t\t\tVHL = 0\n",
    "\t\t\t# 0:'TotalMut', 1: 'TotalNonSyn', 2: 'TotalSNVDriver', 3:'ClonSNVDriver',\n",
    "\t\t\t# 4: 'SubClonSNVDriver', 5: 'DriverSNVRatio', 6: 'ClonSNVRatio'\n",
    "\t\t\t\n",
    "\t\t\t## Calculate CCF for each variant and write to outfile\n",
    "\t\t\tfor line in vcf:\n",
    "\t\t\t\tchrm = line[0]\n",
    "\t\t\t\tchrm = string.split(chrm, 'chr')[1] # removes the 'chr' prefix, if present\n",
    "\t\t\t\tpos = int(line[1])\n",
    "\t\t\t\tfilt = line[6]\n",
    "\t\t\t\tmut_type = string.split(line[7], '|')[1]\n",
    "\t\n",
    "\t\t\t\tif filt == 'PASS':\n",
    "\t\t\t\t\ttmp[0] += 1\n",
    "\t\t\t\tif filt == 'PASS' and mut_type in nonsyn_mut:\n",
    "\t\t\t\t\ttmp[1] += 1\n",
    "\t\t\t\t\t# get mutant and wild-type allele read count for the tumor\n",
    "\t\t\t\t\t# tumor should be last column in VCF\n",
    "\t\t\t\t\tmreads = float(string.split((string.split(line[-1],':')[1]),',')[1])\n",
    "\t\t\t\t\twtreads = float(string.split((string.split(line[-1],':')[1]),',')[0])\n",
    "\t\t\t\t\tdepth = int(mreads + wtreads)\n",
    "\t\t\t\t\tgene = string.split(line[7], '|')[3]\n",
    "\t\t\t\t\tif gene == 'VHL':\n",
    "\t\t\t\t\t\tVHL = 1\n",
    "\t\t\t\t\tif gene in DRIVERS:\n",
    "\t\t\t\t\t\ttmp[2] += 1\n",
    "\t\t\t\t\t\t# Determine chromosomal copy number for normal cells\n",
    "\t\t\t\t\t\t# Autosome copy number = 2\n",
    "\t\t\t\t\t\t# X copy number = 1 if male, or X = 2 if female\n",
    "\t\t\t\t\t\t# Y copy number = 1\n",
    "\t\t\t\t\t\tif chrm == 'X':\n",
    "\t\t\t\t\t\t\tif gender == 0:\n",
    "\t\t\t\t\t\t\t\tcnN = 1.0\n",
    "\t\t\t\t\t\t\telif 'Y' in bic_seg.keys():\n",
    "\t\t\t\t\t\t\t\tcnN = 1.0\n",
    "\t\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\t\tcnN = 2.0\n",
    "\t\t\t\t\t\telif chrm == 'Y':\n",
    "\t\t\t\t\t\t\tcnN = 1.0\n",
    "\t\t\t\t\t\telse:\n",
    "\t\t\t\t\t\t\tcnN = 2.0\n",
    "\t\t\t\t\t\tcnT = Copy_Num(bic_seg, chrm, pos, cnN)\n",
    "\t\t\t\t\t\tccf = CCF(mreads, wtreads, args.pur, cnT, cnN)\n",
    "\t\t\t\t\t\terr = Bootstrap_CCF(mreads, wtreads, args.pur, cnT, cnN)\n",
    "\t\t\t\t\t\tif (ccf + err) >= 1.0: #clonal\n",
    "\t\t\t\t\t\t\ttmp[3] += 1\n",
    "\t\t\t\t\t\telse: #subclona\n",
    "\t\t\t\t\t\t\ttmp[4] += 1\n",
    "\t\t\ttmp[5] = float(tmp[2]/tmp[1])\n",
    "\t\t\ttmp[6] = float(tmp[3]/tmp[2])\n",
    "\t\t\tfor x in tmp:\n",
    "\t\t\t\tFEATURES[t_id].append(x)\n",
    "\t\t\toutfile.write('%s' % (t_id))\n",
    "\t\t\toutfile.write('\\t'.join(FEATURES[t_id]))\n",
    "\t\t\toutfile.write('\\n')\t\t\t\t\n",
    "\t#outfile.write('%s\\t%d\\t%i\\t%f\\t%f\\t%c\\n' % (chrm, pos, depth, ccf, err, clonal))\n",
    "\toutfile.close()\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\tmain()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions for Implementing Linear Discriminate Analysis (LDA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Created on Tue Jun 14 08:36:08 2016\n",
    "\n",
    "@author: mitcheas@ohsum01.ohsu.edu\n",
    "\"\"\"\n",
    "\n",
    "import numpy as np\n",
    "import pandas\n",
    "import seaborn as sns\n",
    "import random\n",
    "#import math\n",
    "#import os, sys, string\n",
    "#import matplotlib.pyplot as plt\n",
    "\n",
    "# LOOCV\n",
    "def LOOCV(X, y):\n",
    "    rows = X.shape[0]\n",
    "    cols = X.shape[1]\n",
    "    E = np.matrix(np.zeros(rows))\n",
    "\n",
    "def TrainLDA(X, y):\n",
    "    # Compute class mean\n",
    "    mu_class = MeanVector(X, y) \n",
    "    \n",
    "    # Compute within class scatter matrix\n",
    "    S_w = WithinClScatter(X, y, mu_class)\n",
    "    \n",
    "    # Compute between class scatter matrix\n",
    "    S_b = BtwnClScatter(X, y, mu_class)\n",
    "    \n",
    "    # Solve eigenvectors\n",
    "    eig_vals, eig_vecs = np.linalg.eig((np.linalg.inv(S_w) * S_b))\n",
    "    print eig_vecs.shape\n",
    "    \n",
    "    # Get new feature subspace\n",
    "    W = subspace(eig_vals, eig_vecs)\n",
    "    \n",
    "    return W\n",
    "\n",
    "    \n",
    "def WithinClScatter(X, y, muvec):\n",
    "    cols = X.shape[1]\n",
    "    numcl = muvec.shape[0]\n",
    "    S = np.zeros((cols, cols))\n",
    "    \n",
    "    for cl, mu in zip(range(0, numcl), muvec):        \n",
    "        X_cl = X[np.where(y == cl)]\n",
    "        w = np.sum((X_cl - mu), axis=0)\n",
    "        w = w.reshape(cols,1)\n",
    "        S += w * w.T       \n",
    "    return S\n",
    "\n",
    "def BtwnClScatter(X, y, muvec):\n",
    "    cols = X.shape[1]\n",
    "    numcl = muvec.shape[0]\n",
    "    S = np.zeros((cols, cols))\n",
    "    \n",
    "    # Compute overall mean\n",
    "    mu_all = np.mean(X, axis=0) \n",
    "    \n",
    "    for cl, mu in zip(range(0, numcl), muvec):\n",
    "        N = X[np.where(y==cl)].shape[0]\n",
    "        mu = mu.reshape(cols, 1)        \n",
    "        mu_all = mu_all.reshape(cols, 1)\n",
    "        S += N * ((mu - mu_all) * (mu - mu_all).T)\n",
    "    return S\n",
    "\n",
    "# Computes and returns the class mean for each feature \n",
    "# X = feature matrix\n",
    "# y = class vector\n",
    "def MeanVector(X, y):\n",
    "    cols = X.shape[1]\n",
    "    c = np.unique(y)\n",
    "    numcl = c.shape[0]\n",
    "    muvec = np.zeros((numcl, cols))\n",
    "    for cl in c:\n",
    "        muvec[cl] = np.mean(X[np.where(y == cl)], axis=0)\n",
    "    return muvec\n",
    "\n",
    "def subspace(eig_vals, eig_vecs, n=2):\n",
    "    cols = eig_vecs.shape[1]\n",
    "    W = np.zeros((n, cols))\n",
    "    \n",
    "    eig_pairs = []\n",
    "    for i in range(0, cols):\n",
    "        eig_pairs.append((np.abs(eig_vals[i]), eig_vecs[:,i]))\n",
    "    \n",
    "    # Sort eigenvectors by eigenvalues in descending order\n",
    "    eig_pairs = sorted(eig_pairs, key=lambda k: k[0], reverse=True)\n",
    "    for i in eig_pairs:\n",
    "        print i\n",
    "   \n",
    "    # Select n number of eigenvectors and store in W   \n",
    "    for i in range(0, n):\n",
    "        W[i] = eig_pairs[i][1]\n",
    "    return W\n",
    "\n",
    "def ProjectLDA(X, W):\n",
    "    rows = X.shape[0]\n",
    "    cols = X.shape[1]\n",
    "    k = W.shape[0]\n",
    "    Y_lda = np.zeros((rows, k))\n",
    "\n",
    "    # Project observations onto new feature subspace\n",
    "    for i in range(0, k):\n",
    "        print W[i].shape \n",
    "        print X.shape\n",
    "        Y_lda[:,i] = X.dot(W[i])\n",
    "    return Y_lda\n",
    "\n",
    "# c = number of classes       \n",
    "def Classify(X, pri, mu_class, W):\n",
    "    c = mu_class.shape[0]\n",
    "    rows = X.shape[0]\n",
    "    cols = X.shape[1]\n",
    "    f = np.zeros((rows, c))\n",
    "    pred = np.zeros((rows))\n",
    "    \n",
    "    # Discriminant function\n",
    "    for cl in range(0, c):\n",
    "        mu = mu_class[cl].reshape(1, cols)\n",
    "        w = W[cl].reshape(1, cols)\n",
    "        a = (mu * w).dot(X.T)\n",
    "        b = 0.5 * ((w * mu).dot(mu.T))\n",
    "        c = np.log2(pri[cl])\n",
    "        f[:, cl] = a - b + c\n",
    "        \n",
    "    # Choose class based on largest discriminant\n",
    "    for i, obs in enumerate(X):\n",
    "        q = f[i]\n",
    "        cl = np.where(q == q.max())\n",
    "        pred[i] = cl[0]\n",
    "        \n",
    "    return pred\n",
    "     \n",
    "def Priors(y):\n",
    "    c = np.unique(y)\n",
    "    numcl = c.shape[0]\n",
    "    n = float(y.shape[0]) \n",
    "    p = np.zeros(numcl)\n",
    "    \n",
    "    for i in c:\n",
    "        p[i] = float(len((np.where(y == i))[0]) / n)\n",
    "    return p\n",
    "    \n",
    "# FUNCTION: ConfMtrx(Y, y)\n",
    "# Calculates the mean squared error of the actual target value (y)\n",
    "# and the predicted target value \n",
    "# y_pred is an array of predicted y values\n",
    "# y_obs is an array of actual class values \n",
    "def ConfMtrx(pred, obs):\n",
    "    err = np.zeros((2,2))\n",
    "    rows = pred.shape[0]\n",
    "    i = 0\n",
    "    while (i < rows):\n",
    "        if obs[i] == pred[i]:\n",
    "            if obs[i] == 1.0:\n",
    "                err[0,0] += 1\n",
    "            else:\n",
    "                 err[1,1] += 1\n",
    "        else:\n",
    "            if obs[i] == 1.0:\n",
    "                err[0,1] += 1\n",
    "            else:\n",
    "                 err[1,0] += 1\n",
    "        i += 1\n",
    "    return err\n",
    "    \n",
    "def MissClassErr(err):\n",
    "    total = float(np.sum((np.sum(err, axis=0)), axis=0))\n",
    "    tp = float(np.trace(err))\n",
    "    fp = float(np.trace(np.fliplr(err)))\n",
    "    e = fp/total\n",
    "    a = tp/total\n",
    "    return e, a"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Implementing LDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "headers = ['GRADE','GENDER','AGE','MAXDIM','METS','VHL','STAGE','TOTALMUT',\n",
    "           'TOTALNONSYN','TOTALDRIVER','CLONDRIVER','SUBCLONDRIVER',\n",
    "           'DRIVRATIO','CLONRATIO']\n",
    "#F = (3,4,6,7,9,12,13)\n",
    "#F = (1,2,3,4,6,10) ## Slected Features\n",
    "F = (1,2,3,4,6,7,8,10,11,12,13) ## All Features\n",
    "np.random.seed(1121)\n",
    "\n",
    "alldata = np.loadtxt('features_KIRC.txt')\n",
    "\n",
    "alldata = random.sample(alldata, len(alldata))\n",
    "alldata = np.asarray(alldata)\n",
    "Xdata = alldata[:,F]\n",
    "#Xdata = alldata[:,(1,2,3,4,6,10)] \n",
    "Ydata = alldata[:,0]\n",
    "\n",
    "#alldata = np.loadtxt('Iris/iris.data', delimiter=',')\n",
    "#t = np.where(alldata[:,-1] != 2)\n",
    "#twoclass = alldata[t]\n",
    "#Xdata = twoclass[:,(0,1,2,3)]\n",
    "#Ydata = twoclass[:,(4)]\n",
    "\n",
    "R = 208\n",
    "Xtrain = Xdata[0:R,]\n",
    "Xtest = Xdata[R:,]\n",
    "Ytrain = Ydata[0:R,]\n",
    "Ytest = Ydata[R:,]\n",
    "\n",
    "Wtrain = TrainLDA(Xtrain, Ytrain)\n",
    "KIRC_lda = ProjectLDA(Xtrain, Wtrain)\n",
    "pri = Priors(Ytrain)\n",
    "mu_class = MeanVector(Xtrain, Ytrain)\n",
    "pred = Classify(Xtest, pri, mu_class, Wtrain)\n",
    "\n",
    "err = ConfMtrx(pred, Ytest)\n",
    "e, a = MissClassErr(err)\n",
    "print \"Error: \", e, \" Accuracy: \", a\n",
    "\n",
    "Y = Ytrain.reshape(R,1)\n",
    "a = np.concatenate((KIRC_lda, Y), axis=1)\n",
    "al = pandas.DataFrame(data=a, columns=('x','y','c'))\n",
    "sns.lmplot('x', 'y', data=al, fit_reg=False, scatter_kws={\"marker\": \"D\", \"s\": 100}, hue='c')\n",
    "plt.show()\n",
    "\n",
    "plt.hist(al.loc[al['c'] == 0]['x'], bins=10, color='b', label='0', alpha=0.7)\n",
    "#plt.hold(True)\n",
    "plt.hist(al.loc[al['c'] == 1]['x'], bins=10, color='g', label='1', alpha=0.7)\n",
    "plt.legend(loc='upper right')\n",
    "plt.show()\n",
    "for i in F:\n",
    "    print headers[i]\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
